Improvement/cldsrv 520 quota monitoring

williamlardier opened the pull request:
Adding monitoring for Quotas evaluation logic

![image](https://github.com/scality/cloudserver/assets/91462779/f6f8382d-30fc-4d3d-a9ee-029b331d9236)

- Health of the scuba service as seen by cloudserver, displayed as a gauge and a % value.
  - Chose to use a counter with two labels in order to ignore when scuba is not deployed
- Number of quota success or failure, displayed as op/s
- The latency, with granularity to know the impact of the quota evaluation logic given different cases: bucket or account quota exceeded, passing, or data deletion.
- The number of bucket with enabled
- The op/s of the quota evaluation by API

An alert (warning) is added if the utilization metrics are stale or not accessible.

BourgoisMickael commented on lib/api/bucketUpdateQuota.js:
```diff
@@ -67,6 +67,7 @@ function bucketUpdateQuota(authInfo, request, log, callback) {
                 'updateBucketQuota');
             return callback(err, err.code, corsHeaders);
         } else {
+            monitoring.numberOfBucketQuota.inc();
```
But cloudserver is stateless, so once it restarts you lose the count of buckets with quota and the panel in the dashboard becomes wrong right ?

williamlardier replied:
The count will stay forever, even if you restart the service (we already follow this logic to count the number of objects)
![image](https://github.com/scality/cloudserver/assets/91462779/3c1f511a-185e-4ebc-ae23-bf9e571b12d0)

To illustrate, I restarted the service after creating the bucket

BourgoisMickael replied:
Yes because it got scraped by prometheus.
But now let's stay you got 100 buckets created in a few seconds and before prometheus can scrape (every 30s on artesca) the server restarts, then the 100 increment is lost.

Also the metric should restart from 0 so the counter will go down if prometheus scrape the metric with a value lower than the one before the restart I think

williamlardier replied:
In this case the information is lost. What I have in mind, and this is also already done today, is to extend the `reportHandler` (called once every hour) to recompute the up to date metrics, along with a new method in arsenal to count the number of bucket (and accounts) with quota enabled. Then we keep this code as-is, and I add a description that explains the metrics are periodically re-computed.

BourgoisMickael replied:
You mean the `reportHandler` will list all buckets to know the up to date number of buckets with quota, number of objects, etc ?

It seems to me this kind of (stateful) counters should maybe not be coming from prometheus but another source like logs form loki

williamlardier replied:
Yes, the report handler lists all buckets every hours and compute various statistics.
The loki source cannot be used as loki is currently disabled in Artesca, so this is not reliable.

BourgoisMickael replied:
Ok so your metric should eventualy be up to date again every hours if there are some restarts in between.

But the list doesn't contain all the information, you'll still have to send another request for each bucket to get their config ?
And this seems quit expensive if you have millions of buckets

williamlardier replied:
Done in https://github.com/scality/cloudserver/pull/5556/commits/24c9a3ca6df345acb3aaa509f8ee69eae2266805
Arsenal was also updated to increment a counter when the processed bucket has a quota set. See https://github.com/scality/Arsenal/pull/2233/commits/63cc54d3fa3c859773cabeef840fe8aad9e3ccae

williamlardier replied:
> But the list doesn't contain all the information, you'll still have to send another request for each bucket to get their config ?

It's an existing code already in prod, we iterate over all the buckets. In a zenko context the number is definitely not millions, but thousands maximum. Anyway millions is also supported, just slower. The change only add a counter, no complexity added.

BourgoisMickael commented on lib/utilities/monitoringHandler.js:
```diff
@@ -64,6 +64,31 @@ const httpResponseSizeBytes = new client.Summary({
     help: 'Cloudserver HTTP response sizes in bytes',
 });
 
+const quotaEvaluationDuration = new client.Histogram({
```
could be nice to have a condition to disable those metrics for S3C if it doesn't need this metrics

williamlardier replied:
I think we will want to have this feature in S3C. For now I am only targeting Zenko branches. We have quotas (account only, and limited) in S3C today though.

francoisferrand replied:
what is the point of disabling the metrics?
if the metric is not "used" in the code (e.g. never `observed()`), no time series will be generated...

BourgoisMickael replied:
Yes, for metrics that have labels like this one, no time series is generated, but for other metrics without label (like `s3_cloudserver_quota_buckets_count`) they are initialized at 0 and will return 0 at every scrape even if unused, so prometheus will keep storing 0 (every 15s on the ring) and increase storage size.

As the prometheus/thanos DB can take a lot of space, it would be nice to not scrape unused metrics for the feature that's not enabled.

williamlardier replied:
Is this approach answering your concerns? https://github.com/scality/cloudserver/pull/5556/files#diff-3daa33d35f5375a7070d74a0b174c01ab8b982392db4025ebc49cf3690aa16bcR75

BourgoisMickael replied:
Yep exactly. And we could do the same in scuba repo for mongo specific metrics

BourgoisMickael commented on lib/utilities/monitoringHandler.js:
```diff
@@ -64,6 +64,31 @@ const httpResponseSizeBytes = new client.Summary({
 });
 
+const quotaEvaluationDuration = new client.Histogram({
+    name: 's3_quota_evaluation_duration_seconds',
```
maybe all the names should start with `s3_cloudserver_`

williamlardier replied:
Yes, I was unsure at that time if we wanted to create a single dashboard for quotas; now that all the logic is in S3, we can keep the `s3_cloudserver_` prefix

BourgoisMickael commented on monitoring/dashboard.py:
```diff
@@ -525,6 +542,111 @@ def top10_errors_by_bucket(title, code):
+        expr='sum by (code) (rate(s3_quota_evaluation_duration_seconds_count{namespace="${namespace}", code=~"2.."}[$__rate_interval]))',  # noqa: E501
+        legendFormat="Success",
+    ), Target(
+        expr='sum by (code) (rate(s3_quota_evaluation_duration_seconds_count{namespace="${namespace}", code=~"4.."}[$__rate_interval]))',  # noqa: E501
+        legendFormat="Quota Exceeded",
```
There is no specific code for quota exceeded, all the errors are quota exceeded ?

What if you receives 400 or 403 or 404 ?

And you don't want to monitor 500 ?

williamlardier replied:
QuotaExceeded is 429

We can create two other filters: for the 4.. (not 429) as bad requests, and 5.. as internal service errors. What do you think?

BourgoisMickael replied:
Yes I think it could be nice to see different errors

williamlardier replied:
Done here: https://github.com/scality/cloudserver/pull/5556/commits/20d2f8bacce5ca1e4b18bbee9747fd8b6a8e88f0#diff-9534a9dab1150e97d07f2e6a3541659b7d37c0aa2fb7db06e95f25fc3864078fR593
We have the success, quota exceeded, bad requests & service errors

BourgoisMickael replied:
I wonder if it's better to call them 4xx and 5xx errors, to not mistake a potential 403 for a 400 bad request

BourgoisMickael commented on monitoring/dashboard.py:
```diff
@@ -525,6 +542,111 @@ def top10_errors_by_bucket(title, code):
+        expr="\n".join([
+            'sum(rate(s3_utilization_service_state{namespace="${namespace}", state="healthy"}[$__rate_interval])) * 100',  # noqa: E501
+            "   /",
+            'sum(rate(s3_utilization_service_state{namespace="${namespace}", state=~"^(unhealthy|healthy)$"}[$__rate_interval]))',  # noqa: E501
```
For promql I think you don't need `^$` in regex:

> Regex matches are fully anchored. A match of env=\~"foo" is treated as env=~"^foo$".

https://prometheus.io/docs/prometheus/latest/querying/basics/

```suggestion
            'sum(rate(s3_utilization_service_state{namespace="${namespace}", state=~"unhealthy|healthy"}[$__rate_interval]))',  # noqa: E501

```

BourgoisMickael commented on monitoring/dashboard.py:
```diff
@@ -525,6 +542,111 @@ def top10_errors_by_bucket(title, code):
+            title="Account quota exceeded", type='"bucketAccount"',
+            code='~"4.."'),
+        average_quota_latency_target(
+            title="Quota exceeded (delete data)", type='"delete"'),
```
`data` is a little vague, is it all delete action ?
Also is the delete included in the above quota on account and bucket ?

williamlardier replied:
This is indeed for the delete APIs. I can rename it "Data deletion".
This metric is not part of the above values, it's a special case when you delete data, the other ones track the writes.

BourgoisMickael replied:
maybe you should describe that as well for other metrics

williamlardier replied:
I reworked the timeserie, now we say if we write or delete, with some details:

![image](https://github.com/scality/cloudserver/assets/91462779/eaf4c4b7-9a35-4f7b-8294-18e552b39acd)

BourgoisMickael replied:
for `Data write (bucket)` you could add that it's for success so we don't think it encapsulate the exceeded also

BourgoisMickael commented on monitoring/dashboard.py:
```diff
@@ -525,6 +542,111 @@ def top10_errors_by_bucket(title, code):
+    noValue="0",
+    reduceCalc="lastNotNull",
+    targets=[Target(
+        expr='sum(s3_cloudserver_buckets_quota_count{namespace="${namespace}"})',  # noqa: E501
```
```suggestion
        expr='sum(s3_quota_evaluation_duration_seconds_count{namespace="${namespace}"})',  # noqa: E501
```

BourgoisMickael replied:
No nevermind, I wrote the wrong thing I think

BourgoisMickael commented on monitoring/dashboard.py:
```diff
@@ -525,6 +542,111 @@ def top10_errors_by_bucket(title, code):
 )
 
+
+quotaHealth = GaugePanel(
+    title="Quota service uptime",
```
It's a little weird to show a gauge for the `mean` during `__rate_interval`

You could see more info if you graph in a time series to see over time when it was up or down and also know at the current time if it's up or down without a `mean`

williamlardier replied:
Updated with on/off display over time:

![image](https://github.com/scality/cloudserver/assets/91462779/a767538f-19b2-48ec-bfc9-23bf527d7b08)

Not related, but FYI, I also added another graph, with the utilization metrics retrieval latencies, I was not using the metric before, while it's useful to see the impact of the internal calls in during the quota evaluation logic:

![image](https://github.com/scality/cloudserver/assets/91462779/99c7c7de-9fb5-4eee-93f2-411a0617ec21)
(the one in the middle, bottom)

BourgoisMickael replied:
About On / Off, I think you have a percentage with your promql, so you could keep displaying a percentage to see if you have some instances that can't connect or scuba down over time rather than a boolean

BourgoisMickael replied:
And for the metric retrieval latencies, should you really have a big diff between this and the quota evaluation ? Also you only record success for metric retriaval

williamlardier replied:
Yes, metric retrieval can be done twice in an API call, in case we need to cleanup when quota is exceeded. We still want to be able the latency with scuba, independently from the quota evaluation logic

(please feel free to challenge that)

BourgoisMickael replied:
I think the utilization metric retrieval metric is the same as `scuba_http_request_duration_seconds{action="GetMetrics"}`
(https://github.com/scality/scuba/pull/104/files#diff-7c16f2a12075d20e3640e31bcb11cc1b87aad07a0fc048ace705020732a5a2cfR21-R28) but with network time over it.

But then maybe scuba metrics can record request that come from other places than cloudserver, so maybe you should actually keep this

___
Otherwise I posted in another comment, it could be nice to have another scuba endpoint to aggregate all the requests into one, especially if it's a code path that's called often.

williamlardier replied:
> But then maybe scuba metrics can record request that come from other places than cloudserver, so maybe you should actually keep this

Yes it's indeed the same, but this is useful to see the impact of internal network communications, as seen by S3, as the scuba metrics may be, in the future, including public ones with authz, expected to be slower...

For the other endpoint, yes, see https://scality.atlassian.net/browse/ZENKO-4801 (the current epic supports account, but the use case is bucket)

BourgoisMickael commented on lib/api/apiUtils/authorization/permissionChecks.js:
```diff
@@ -396,6 +399,12 @@ async function validateQuotas(bucket, account, apiNames, apiMethod, inflight, lo
                 bucketName: bucket.getName(),
             });
         }
+        const responseTimeInNs = Number(process.hrtime.bigint() - requestStartTime);
+        monitoring.quotaEvaluationDuration.labels({
+            action: apiMethod,
+            type: 'noCheck',
+            code: 200,
+        }).observe(responseTimeInNs / 1e9);
```
I don't see the `noCheck` in the dashboard, and I'm not sure this one provides any value, if there is no quota evaluation it could be ok to not monitor anything to not have unused metric, except if you plan to use it later ?

williamlardier replied:
Yes, I agree that there is no point in monitoring "nothing", I will remove it

williamlardier replied:
Removed in https://github.com/scality/cloudserver/pull/5556/commits/9facef770aed6b5a13ad72993095f74d20f7eb09

BourgoisMickael commented on lib/api/apiUtils/authorization/permissionChecks.js:
```diff
@@ -453,30 +462,62 @@ async function validateQuotas(bucket, account, apiNames, apiMethod, inflight, lo
+        if (actionWithDataDeletion[apiMethod]) {
+            const responseTimeInNs = Number(process.hrtime.bigint() - requestStartTime);
+            monitoring.quotaEvaluationDuration.labels({
+                action: apiMethod,
```
as pointed here https://github.com/scality/Arsenal/pull/2233/files#r1568992349 I think if `apiMethod` comes from request, you might have the delete object version being called `objectDelete`

williamlardier replied:
Yes, this is fine, we do not need to differentiate between versioned and non-versioned deletes

BourgoisMickael commented on lib/api/apiUtils/authorization/permissionChecks.js:
```diff
@@ -453,30 +462,62 @@ async function validateQuotas(bucket, account, apiNames, apiMethod, inflight, lo
                 // eslint-disable-next-line no-await-in-loop
-                await ScubaClientInstance.getLatestMetrics('bucket',
+                await ScubaClientInstance.getUtilizationMetrics('bucket',
                     `${bucket.getName()}_${creationDate}`, null, {
                     action: apiMethod,
                     inflight: -inflight,
                 });
                 if (account?.quota) {
                     // eslint-disable-next-line no-await-in-loop
-                    await ScubaClientInstance.getLatestMetrics('account',
+                    await ScubaClientInstance.getUtilizationMetrics('account',
```
would it be more interesting to have another endpoint in scuba to retrieve both bucket and account metrics so you send only one request ?

BourgoisMickael replied:
and also where you could pass both `inflight` and `-inflight` for delete to reduce the number of api calls

williamlardier replied:
> would it be more interesting to have another endpoint in scuba to retrieve both bucket and account metrics so you send only one request ?

Yes, this is something we can do; for now the use case will be to have only bucket quotas, but in case they have both, we can add another API

But this breaks compatibility with the existing scuba APIs, so better doing that in a future iteration to improve the performance, what do you think?

BourgoisMickael replied:
yep, maybe make a ticket so we don't forget

williamlardier replied:
I created an epic and the ticket is inside: https://scality.atlassian.net/browse/ARTESCA-12124
We already identified a bunch of things we need to improve, so this epic will be useful for that.

BourgoisMickael commented on lib/api/apiUtils/authorization/permissionChecks.js:
```diff
@@ -453,30 +462,62 @@ async function validateQuotas(bucket, account, apiNames, apiMethod, inflight, lo
+        }).observe(responseTimeInNs / 1e9);
+
         return callback();
     } catch (err) {
```
Should you observe the quota evaluation duration in the catch for errors from Scuba ? I don't see how you got some Bad Requests in your graph otherwise

williamlardier replied:
Done in https://github.com/scality/cloudserver/pull/5556/commits/0006a0f07d83cb818352e0f6caec6662323aee19

I chose the error code 202, because in case of error we still process the request. Note that the error from scubaclient is also monitored, so information is not lost.

BourgoisMickael commented on lib/scuba/wrapper.js:
```diff
@@ -36,6 +40,9 @@ class ScubaClientImpl extends ScubaClient {
             }
+            monitoring.utilizationServiceHealth.labels({
+                state: 'unhealthy',
+            }).inc();
```
Incrementing counter is not ideal here if you want to have an accurate healthy percentage in your panel.
If scuba goes up and down multiple times, you'll have counters that keep incrementing.

You might want to use a gauge to set a boolean like this:
```js
// success
utilizationServiceHealth.labels(healthy).set(1)
utilizationServiceHealth.labels(unhealthy).set(0)
// error
utilizationServiceHealth.labels(healthy).set(0)
utilizationServiceHealth.labels(unhealthy).set(1)
```

williamlardier replied:
Done in https://github.com/scality/cloudserver/pull/5556/commits/8ef63ab5cd687bbe34c0a5eb42f8d56821a641e2

BourgoisMickael commented on lib/scuba/wrapper.js:
```diff
@@ -50,6 +57,20 @@ class ScubaClientImpl extends ScubaClient {
+        return this.getLatestMetrics(metricsClass, resourceName, options, body).then(data => {
+            const responseTimeInNs = Number(process.hrtime.bigint() - requestStartTime);
+            monitoring.utilizationMetricsRetrievalDuration.labels({
+                method: 'POST',
+                action: 'getUtilizationMetrics',
+                code: 200,
+                class: metricsClass,
```
Should you also monitor errors ? Otherwise the `code` could be removed, but your dashboard is ready to accept errors on this metric.

Constant `method` and `action` labels here seems unnecessary and could be removed.

`class` is useful ? Should you display it on your dashboard ?

williamlardier replied:
I will remove method and action indeed.

The `class` is useful because I display the latency for the following cases:
- bucket, success
- bucket, error
- account, success
- account, error

Later we will also have both, if we mutualize in a single API

williamlardier replied:
Done in https://github.com/scality/cloudserver/pull/5556/commits/460404bd7391f06347009eb2e5bd4063db5cccca

BourgoisMickael commented on monitoring/alerts.yaml:
```diff
@@ -132,3 +135,17 @@ groups:
+  - alert: UtilizationMetricsServiceNotAvailable
+    expr: |
+      sum(rate(s3_cloudserver_utilization_service_state{namespace="${namespace}",state="healthy"}[1m])) * 100
+          / sum(rate(s3_cloudserver_utilization_service_state{namespace="${namespace}",state=~"^(unhealthy|healthy)$"}[1m]))
```
```suggestion
          / sum(rate(s3_cloudserver_utilization_service_state{namespace="${namespace}",state=~"unhealthy|healthy"}[1m]))
```

same as the dashboard one

williamlardier replied:
Done in https://github.com/scality/cloudserver/pull/5556/commits/95ef7fcd9db20f1ae3b11e5590c27e7a76fc45cf

BourgoisMickael commented on lib/utilities/monitoringHandler.js:
```diff
@@ -64,6 +64,31 @@ const httpResponseSizeBytes = new client.Summary({
+});
+
+const utilizationServiceHealth = new client.Counter({
+    name: 's3_cloudserver_utilization_service_state',
```
why not call it scuba ?

williamlardier replied:
I don't know if we want to use Scuba publicly, because it was supposed to be dropped in favor of SUR, but anyway the reason is that I wanted to dissociate the naming of the metric with the backend, as the backend can change independently of cloudserver...

francoisferrand replied:
`utilization_service_state` seems out of place (the state of the utilization service should be monitored in that service... here we monitor the availability of the feature in cloudserver which depends on it, i.e. the impact of unavailability of that service?), and it feels like we loose the "relationship" with the quota.

maybe something like `s3_cloudserver_quota_availaibility` or `s3_cloudserver_quota_state` ?

williamlardier replied:
Done in https://github.com/scality/cloudserver/pull/5556/files#diff-3daa33d35f5375a7070d74a0b174c01ab8b982392db4025ebc49cf3690aa16bcR91

I still have the utilization service inside as this is what defines the availability of the feature

francoisferrand commented on monitoring/alerts.yaml:
```diff
@@ -132,3 +135,17 @@ groups:
+  # but not available during the last 5 minutes
+  - alert: UtilizationMetricsServiceNotAvailable
+    expr: |
+      sum(rate(s3_cloudserver_utilization_service_state{namespace="${namespace}",state="healthy"}[1m])) * 100
+          / sum(rate(s3_cloudserver_utilization_service_state{namespace="${namespace}",state=~"unhealthy|healthy"}[1m]))
+        < ${utilizationServiceUnavailabilityThreshold}
```
should filter on `${service}`

francoisferrand commented on monitoring/dashboard.py:
```diff
@@ -366,6 +366,38 @@ def average_latency_target(title, action="", by=""):
+    by = " by (" + by + ")" if by else ""
+    return Target(
+        expr="\n".join([
+            'sum(rate(s3_cloudserver_quota_evaluation_duration_seconds_sum{namespace="${namespace}"' + extra + "}[$__rate_interval]))" + by,  # noqa: E501
+            "   /",
+            'sum(rate(s3_cloudserver_quota_evaluation_duration_seconds_count{namespace="${namespace}"' + extra + "}[$__rate_interval]))" + by,  # noqa: E501,
```
need to filter on service/job

francoisferrand commented on monitoring/dashboard.py:
```diff
@@ -366,6 +366,38 @@ def average_latency_target(title, action="", by=""):
+    by = " by (" + by + ")" if by else ""
+    return Target(
+        expr="\n".join([
+            'sum(rate(s3_cloudserver_utilization_metrics_retrieval_duration_seconds_sum{namespace="${namespace}"' + extra + "}[$__rate_interval]))" + by,  # noqa: E501
+            "   /",
+            'sum(rate(s3_cloudserver_utilization_metrics_retrieval_duration_seconds_count{namespace="${namespace}"' + extra + "}[$__rate_interval]))" + by,  # noqa: E501,
```
need to filter on service/job

francoisferrand commented on lib/utilities/monitoringHandler.js:
```diff
@@ -64,6 +64,36 @@ const httpResponseSizeBytes = new client.Summary({
+const numberOfAccountQuota = new client.Gauge({
+    name: 's3_cloudserver_account_quota_count',
+    help: 'Total number of account quota',
+});
```
`s3_cloudserver_quota_buckets_count` and `s3_cloudserver_quota_accounts_count`

francoisferrand commented on lib/utilities/monitoringHandler.js:
```diff
@@ -64,6 +64,36 @@ const httpResponseSizeBytes = new client.Summary({
 });
 
+const quotaEvaluationDuration = new client.Histogram({
+    name: 's3_cloudserver_quota_evaluation_duration_seconds',
+    help: 'Duration of the quota evaluation operation',
+    labelNames: ['action', 'code', 'type'],
+    buckets: [0.0001, 0.005, 0.015, 0.05, 0.1, 0.2, 0.3, 0.4, 0.5],
+});
+
+const utilizationMetricsRetrievalDuration = new client.Histogram({
+    name: 's3_cloudserver_utilization_metrics_retrieval_duration_seconds',
```
do we need to differentiate both durations ?
i.e. once the utilization metrics have been read, quota evaluation is just a few ops, with no more network ops or complex computation?

williamlardier replied:
Yes, because we might do more than one call to scuba in a single evaluation loop, for example to remove the inflights after detecting a quota exceeded, this help shaving visibility on the internal call latencies, versus the quota evaluation logic durations (with different cases)

francoisferrand commented on lib/utilities/monitoringHandler.js:
```diff
@@ -64,6 +64,36 @@ const httpResponseSizeBytes = new client.Summary({
 });
 
+const quotaEvaluationDuration = new client.Histogram({
+    name: 's3_cloudserver_quota_evaluation_duration_seconds',
+    help: 'Duration of the quota evaluation operation',
+    labelNames: ['action', 'code', 'type'],
+    buckets: [0.0001, 0.005, 0.015, 0.05, 0.1, 0.2, 0.3, 0.4, 0.5],
+});
+
+const utilizationMetricsRetrievalDuration = new client.Histogram({
+    name: 's3_cloudserver_utilization_metrics_retrieval_duration_seconds',
```
i think we should keep the same "prefix" for all these metrics, to make it easier to manage the (long) list of metrics...

not sure how best to do it in this case (unless it makes sens to remove the metrics, c.f. earlier comment 😁), the only thing I could come up with is `s3_cloudserver_quota_retrieval_duration_seconds` (which is not _correct_, but probably good enough...)

williamlardier replied:
I will use `s3_cloudserver_quota_metrics_retrieval_duration_seconds`

williamlardier replied:
See https://github.com/scality/cloudserver/pull/5556/files#diff-3daa33d35f5375a7070d74a0b174c01ab8b982392db4025ebc49cf3690aa16bcR84

francoisferrand commented on lib/utilities/monitoringHandler.js:
```diff
@@ -64,6 +64,36 @@ const httpResponseSizeBytes = new client.Summary({
+const quotaEvaluationDuration = new client.Histogram({
+    name: 's3_cloudserver_quota_evaluation_duration_seconds',
+    help: 'Duration of the quota evaluation operation',
+    labelNames: ['action', 'code', 'type'],
```
do we need to filter by `code` or `action` ? i got mixed feeling here:
* we don't really care about the action with regard to the _duration_, since all these actions should be affected the same way?
* the code (= result of the evaluation) is definitely not relevant for the _duration_
* however, it could indeed be interesting to have this labels to allow checking which requests cause the most quota checks or the quota check success/rate
* but at the same time, we can probably get this information by just looking at the request count (which has action & code).... the only thing missing from the request count may be the `type` of quota check (i.e. none, bucket, account, bucket+account)

williamlardier replied:
Depending on the API (write vs delete) and the response (quota exceeded vs passing), both action and code are relevant to discriminate between the different flows of the quota evaluation logic
For example quota exceeded will have more impact as we need to clean the inflights in case of quota exceeded

francoisferrand commented on lib/utilities/monitoringHandler.js:
```diff
@@ -64,6 +64,36 @@ const httpResponseSizeBytes = new client.Summary({
+const quotaEvaluationDuration = new client.Histogram({
+    name: 's3_cloudserver_quota_evaluation_duration_seconds',
+    help: 'Duration of the quota evaluation operation',
+    labelNames: ['action', 'code', 'type'],
```
we may have either account or bucket quotas, but could also have both : how is this managed in the `type` label? should we maybe have 2 labels instead (`bucket` and `account`) ?

williamlardier replied:
Right now we have 3 different strings (bucket, account, bucketAccount), if we want to add more types later, I thought it made sense to have a single label for it...

francoisferrand commented on lib/utilities/monitoringHandler.js:
```diff
@@ -64,6 +64,36 @@ const httpResponseSizeBytes = new client.Summary({
+    name: 's3_cloudserver_quota_evaluation_duration_seconds',
+    help: 'Duration of the quota evaluation operation',
+    labelNames: ['action', 'code', 'type'],
+    buckets: [0.0001, 0.005, 0.015, 0.05, 0.1, 0.2, 0.3, 0.4, 0.5],
```
`0.1ms` does not seem realistic...
Also, you mentioned the impact (in your test) was around 1ms, so i guess we should have more buckets in this range: maybe `[1ms, 2,5ms, 5ms, 10ms, 25ms, 50ms, 100ms, 200ms, 400ms, 800ms]` ?

BourgoisMickael replied:
Then you can also change the buckets for the metric_retrieval_duration metric as the quota depends on this one and could be double the metric retrieval time ?

And then maybe change as well the buckets in scuba to match with the retrieval buckets ?

___

Also I wonder if it could be possible that  the time will differ between a mongodb backend for artesca and a leveldb/metadata backend for the ring and we should have different buckets configured for each product ?

williamlardier replied:
Updated in https://github.com/scality/cloudserver/pull/5556/files#diff-3daa33d35f5375a7070d74a0b174c01ab8b982392db4025ebc49cf3690aa16bcR80

I will assess if changes are needed in Scuba, as the latency in scuba is lower (no network latency added, as compared with cloudserver)

francoisferrand commented on lib/utilities/monitoringHandler.js:
```diff
@@ -64,6 +64,36 @@ const httpResponseSizeBytes = new client.Summary({
+const utilizationServiceHealth = new client.Gauge({
+    name: 's3_cloudserver_utilization_service_state',
+    help: 'Availability of the utilization service',
+    labelNames: ['state'],
```
what are the possible states? and values for the metric?

francoisferrand commented on lib/api/apiUtils/authorization/permissionChecks.js:
```diff
@@ -385,10 +387,19 @@ function processBucketPolicy(requestType, bucket, canonicalID, arn, bucketOwner,
     let accountQuotaExceeded = false;
+    let type = 'bucket';
+    if (bucketQuota && accountQuota) {
+        type = 'bucketAccount';
```
probably should make it parsable: `bucket+account` or `bucket,account` ?
(or use separate labels, c.f. other comment)

williamlardier replied:
I will make it parsable indeed, will choose the first option

williamlardier replied:
See https://github.com/scality/cloudserver/pull/5556/files#diff-773344fcfc0f64145ff95eb6eccd556588c19449dc2c263a2532d4030bf00dffR407

francoisferrand commented on lib/api/apiUtils/authorization/permissionChecks.js:
```diff
@@ -385,10 +387,19 @@ function processBucketPolicy(requestType, bucket, canonicalID, arn, bucketOwner,
  * @returns {boolean} - true if the quota is valid, false otherwise
  */
 async function validateQuotas(bucket, account, apiNames, apiMethod, inflight, log, callback) {
```
probably not in this PR, but this function is very weird: being both async and with a callback...

williamlardier replied:
As discussed I will callbackify the call to scubaclient in the PR this one is based on

williamlardier replied:
Callbackify was done in the base PR.

francoisferrand commented on lib/api/apiUtils/authorization/permissionChecks.js:
```diff
@@ -453,39 +462,70 @@ async function validateQuotas(bucket, account, apiNames, apiMethod, inflight, lo
 
+        if (actionWithDataDeletion[apiMethod]) {
+            const responseTimeInNs = Number(process.hrtime.bigint() - requestStartTime);
+            monitoring.quotaEvaluationDuration.labels({
```
can we move this up (where the function is called), to avoid duplication?

williamlardier replied:
No because this is called in a centralized place, and in specific APIs sometimes (copy object and multi object deletes need that, as we do not have the obj metadata at the authorization time)...

francoisferrand commented on lib/scuba/wrapper.js:
```diff
@@ -28,6 +29,8 @@ class ScubaClientImpl extends ScubaClient {
                 this._log.info('Scuba health check passed, enabling quotas');
             }
+            monitoring.utilizationServiceHealth.labels({ state: 'healthy' }).set(1);
+            monitoring.utilizationServiceHealth.labels({ state: 'unhealthy' }).set(0);
```
avoid multiple metrics like this does not seem very "user friendly": a single `available` metric with a pseuod-boolean value (0 = no, 1 = yes), like the `up` metrics, seems a more standard approach...

williamlardier replied:
Now we have no label https://github.com/scality/cloudserver/pull/5556/files#diff-7a879f848f4d512bac1250da6f152d89b8feab003de8d9fc6fa202cb6b17dcf5R34

francoisferrand commented on lib/utilities/monitoringHandler.js:
```diff
@@ -64,6 +64,36 @@ const httpResponseSizeBytes = new client.Summary({
+    labelNames: ['state'],
+});
+
+const numberOfBucketQuota = new client.Gauge({
```
numberOfXXX → XXXXCount

francoisferrand commented on monitoring/dashboard.py:
```diff
@@ -386,6 +418,7 @@ def average_latency_target(title, action="", by=""):
     ],
 )
 
+
```
unrealted

francoisferrand commented on monitoring/dashboard.py:
```diff
@@ -400,6 +433,7 @@ def average_latency_target(title, action="", by=""):
     ]
 )
 
+
```
unrelated

francoisferrand commented on monitoring/dashboard.py:
```diff
@@ -525,6 +559,167 @@ def top10_errors_by_bucket(title, code):
+        expr="\n".join([
+            'sum(rate(s3_cloudserver_utilization_service_state{namespace="${namespace}", state="healthy"}[$__rate_interval])) * 100',  # noqa: E501
+            "   /",
+            'sum(rate(s3_cloudserver_utilization_service_state{namespace="${namespace}", state=~"unhealthy|healthy"}[$__rate_interval]))',  # noqa: E501
```
for the `uptime`, i think it is enough to graph `s3_cloudserver_utilization_service_state{state=healthy}` : it is already 0 when down and 1 when up...

or may you are trying to compute the % of time the service is available _over time_ ?
in that case, it should be more something like `avg_over_time(s3_cloudserver_utilization_service_state{state=healthy})`...

francoisferrand commented on monitoring/dashboard.py:
```diff
@@ -525,6 +559,167 @@ def top10_errors_by_bucket(title, code):
+    unit=UNITS.PERCENT_FORMAT,
+    targets=[Target(
+        expr="\n".join([
+            'sum(rate(s3_cloudserver_utilization_service_state{namespace="${namespace}", state="healthy"}[$__rate_interval])) * 100',  # noqa: E501
```
must not use `rate` on gauge (only on counters)

francoisferrand commented on monitoring/dashboard.py:
```diff
@@ -525,6 +559,167 @@ def top10_errors_by_bucket(title, code):
+    lineInterpolation="smooth",
+    unit=UNITS.OPS_PER_SEC,
+    targets=[Target(
+        expr='sum by (code) (rate(s3_cloudserver_quota_evaluation_duration_seconds_count{namespace="${namespace}", code=~"2.."}[$__rate_interval]))',  # noqa: E501
+        legendFormat="Success",
```
* either you `sum by(code)`, to get one time serie per code → the legendFormat must be based on the code (typically `{{ code }}`)
* or you want a single serie "success" → use a simpler sum, without grouping
* 
```suggestion
        expr='sum(rate(s3_cloudserver_quota_evaluation_duration_seconds_count{namespace="${namespace}", code=~"2.."}[$__rate_interval]))',  # noqa: E501
        legendFormat="Success",
```

francoisferrand replied:
must also filter by job/service

francoisferrand commented on monitoring/dashboard.py:
```diff
@@ -525,6 +559,167 @@ def top10_errors_by_bucket(title, code):
+        expr='sum by (code) (rate(s3_cloudserver_quota_evaluation_duration_seconds_count{namespace="${namespace}", code=~"2.."}[$__rate_interval]))',  # noqa: E501
+        legendFormat="Success",
+    ), Target(
+        expr='sum by (code) (rate(s3_cloudserver_quota_evaluation_duration_seconds_count{namespace="${namespace}", code="429"}[$__rate_interval]))',  # noqa: E501
+        legendFormat="Quota Exceeded",
+    ), Target(
+        expr='sum by (code) (rate(s3_cloudserver_quota_evaluation_duration_seconds_count{namespace="${namespace}", code=~"4..", code!="429"}[$__rate_interval]))',  # noqa: E501
+        legendFormat="Status 4xx",
+    ), Target(
+        expr='sum by (code) (rate(s3_cloudserver_quota_evaluation_duration_seconds_count{namespace="${namespace}", code=~"5.."}[$__rate_interval]))',  # noqa: E501
+        legendFormat="Status 5xx",
```
same

williamlardier replied:
Removed.

francoisferrand commented on monitoring/dashboard.py:
```diff
@@ -525,6 +559,167 @@ def top10_errors_by_bucket(title, code):
+    ), Target(
+        expr='sum by (code) (rate(s3_cloudserver_quota_evaluation_duration_seconds_count{namespace="${namespace}", code="429"}[$__rate_interval]))',  # noqa: E501
+        legendFormat="Quota Exceeded",
+    ), Target(
+        expr='sum by (code) (rate(s3_cloudserver_quota_evaluation_duration_seconds_count{namespace="${namespace}", code=~"4..", code!="429"}[$__rate_interval]))',  # noqa: E501
+        legendFormat="Status 4xx",
+    ), Target(
+        expr='sum by (code) (rate(s3_cloudserver_quota_evaluation_duration_seconds_count{namespace="${namespace}", code=~"5.."}[$__rate_interval]))',  # noqa: E501
+        legendFormat="Status 5xx",
```
are these 4xx and 5xx status relevant for this metric? at the end of the evaluation, we can only have either 'proceed' or '429'... Even if the query eventually ends up with another error code, the metric will have already recorded either 200 or 429...

williamlardier replied:
Removed.

francoisferrand commented on monitoring/dashboard.py:
```diff
@@ -525,6 +559,167 @@ def top10_errors_by_bucket(title, code):
+    noValue="0",
+    reduceCalc="lastNotNull",
+    targets=[Target(
+        expr='sum(s3_cloudserver_account_quota_count{namespace="${namespace}"})',  # noqa: E501
```
`max()` may be more appropriate than `sum`, it case we have multiple series at the same time (e.g. pod update maybe?)

williamlardier replied:
Indeed, see https://github.com/scality/cloudserver/pull/5556/files#diff-9534a9dab1150e97d07f2e6a3541659b7d37c0aa2fb7db06e95f25fc3864078fR696

francoisferrand commented on monitoring/dashboard.py:
```diff
@@ -525,6 +559,167 @@ def top10_errors_by_bucket(title, code):
+    dataSource="${DS_PROMETHEUS}",
+    colorMode="value",
+    format=UNITS.SHORT,
+    noValue="0",
```
if we don't have a value, we may want to make it visibile we don't actually know : instead of reporting a possibly false `0`...

williamlardier replied:
Removed.

francoisferrand commented on monitoring/dashboard.py:
```diff
@@ -525,6 +559,167 @@ def top10_errors_by_bucket(title, code):
+    dataSource="${DS_PROMETHEUS}",
+    colorMode="value",
+    format=UNITS.SHORT,
+    noValue="0",
```
if we don't have a value, we may want to make it visibile we don't actually know : instead of reporting a possibly false 0...

francoisferrand commented on monitoring/dashboard.py:
```diff
@@ -525,6 +559,167 @@ def top10_errors_by_bucket(title, code):
+    noValue="0",
+    reduceCalc="lastNotNull",
+    targets=[Target(
+        expr='sum(s3_cloudserver_buckets_quota_count{namespace="${namespace}"})',  # noqa: E501
```
max() may be more appropriate than sum, it case we have multiple series at the same time (e.g. pod update maybe?)

francoisferrand commented on monitoring/dashboard.py:
```diff
@@ -630,6 +819,18 @@ def top10_errors_by_bucket(title, code):
+                quotaByAction,
+            ], height=8),
+            layout.row([
+                layout.resize([bucketQuotaCounter], width=3),
+                layout.resize([accountQuotaCounter], width=3),
```
we probably want to stack these on top of each other, will look better (they look quite stretched, due to the small width & large height)

francoisferrand commented on monitoring/dashboard.py:
```diff
@@ -630,6 +819,18 @@ def top10_errors_by_bucket(title, code):
+            layout.row([
+                layout.resize([bucketQuotaCounter], width=3),
+                layout.resize([accountQuotaCounter], width=3),
+                layout.resize([averageMetricsRetrievalLatencies], width=6),
+                averageQuotaLatencies,
```
we store histogram metrics, but only display the average → we should try to use the information to provide more info, like the distribution (heatmap) or 95th percentile...

williamlardier replied:
What do you think of the heatmap I added? (see comment above)

BourgoisMickael commented on monitoring/dashboard.py:
```diff
@@ -525,6 +557,173 @@ def top10_errors_by_bucket(title, code):
 )
 
+
+quotaHealth = TimeSeries(
```
you might want to put a max at `100%` as it shows 200% in your screenshot.

In the legend you could put the latest value so we don't have to hover to see the tooltip

williamlardier replied:
Unfortunately it's not yet available in the python lib we use, I'll see if I can do something about it.

BourgoisMickael commented on monitoring/dashboard.py:
```diff
@@ -525,6 +557,173 @@ def top10_errors_by_bucket(title, code):
+    dataFormat="tsbuckets",
+    maxDataPoints=25,
+    tooltip=Tooltip(show=True, showHistogram=True),
+    yAxis=YAxis(format=UNITS.DURATION_SECONDS),
```
I don't know why grafana displays that duration seconds with `milliseconds` instead of `ms` but if you pick only seconds instead of duration seconds, it seems to abbreviate correctly and display `ms`.
But I'm not sure what's the difference between the 2 units

Seragonia committed:
CLDSRV-520: add metrics for quota

Seragonia committed:
CLDSRV-520: observe metrics during quota evaluations

Seragonia committed:
CLDSRV-520: observe number of buckets and accounts with quota

Seragonia committed:
CLDSRV-520: add alerts

Seragonia committed:
CLDSRV-520: add dashboards

francoisferrand commented on monitoring/dashboard.py:
```diff
@@ -525,6 +557,173 @@ def top10_errors_by_bucket(title, code):
+    unit=UNITS.PERCENT_FORMAT,
+    targets=[Target(
+        expr='avg(avg_over_time(s3_cloudserver_quota_utilization_service_available{namespace="${namespace}",job=~"$job"}[$__rate_interval])) * 100',  # noqa: E501
+        legendFormat="State",
```
legend not needed, since there is only one series

williamlardier replied:
Updated

francoisferrand commented on monitoring/dashboard.py:
```diff
@@ -525,6 +557,172 @@ def top10_errors_by_bucket(title, code):
+    yAxis=YAxis(format=UNITS.SECONDS),
+    color=HeatmapColor(mode="opacity"),
+    targets=[Target(
+        expr='sum by(le) (increase(s3_cloudserver_quota_evacdluation_duration_seconds_bucket{namespace="${namespace}", job=~"$job"}[$__rate_interval]))',  # noqa: E501
```
by convention, prefer to use braces around _inputs_ variables (everywhere in this file): makes it easier to see with syntax highlighting (esp. in github) and differentiate from _template_ variables (managed by grafana at runtime).

Also, you have a single job, so you can use the simpler (and likely faster) exact match operator.

```suggestion
        expr='sum by(le) (increase(s3_cloudserver_quota_evacdluation_duration_seconds_bucket{namespace="${namespace}", job="${job}"}[$__rate_interval]))',  # noqa: E501
```

francoisferrand commented on monitoring/dashboard.py:
```diff
@@ -525,6 +557,172 @@ def top10_errors_by_bucket(title, code):
+                                        _class='"bucket"', code='~"4..|5.."'),
+        average_quota_retrieval_latency(title="Account (error)",
+                                        _class='"account"', code='~"4..|5.."'),
+    ],
```
what about `bucket+account` ?

To avoid duplication, you can simply define 2 targets where the prometheus query has a `by(class)` clause, something like (using functions may be fine, but probably cleaner to use them only to build the expression):

```
targets:[
    Target(
        expr: 'sum(rate(quota_metrics_retrieval{code=~"2.."}[$__rate_interval])) by(class)',
        legendFormat: '{{ class }} (success)',
    ),
    Target(
        expr: 'sum(rate(quota_metrics_retrieval{code=~"4..|5.."}[$__rate_interval])) by(class)',
        legendFormat: '{{ class }} (error)',
    )
]
```

williamlardier replied:
Were you thinking of something like this https://github.com/scality/cloudserver/pull/5556/commits/5b7dd3b8413927327636c3fad2c021369bb946be ?
I'll deploy it tomorrow and see if it needs some updates / improvements, and I will also try to display the `s3_cloudserver_quota_request_with_metrics_unavailable` metric.

williamlardier replied:
Tested in my platform with success
![image](https://github.com/scality/cloudserver/assets/91462779/9fea1181-522b-4bd6-8316-dd20722ddd3b)

Also fixed the type for deletes, as it was never set.

francoisferrand commented on monitoring/dashboard.py:
```diff
@@ -525,6 +557,172 @@ def top10_errors_by_bucket(title, code):
+    noValue="-",
+    reduceCalc="lastNotNull",
+    targets=[Target(
+        expr='max(s3_cloudserver_quota_buckets_count{namespace="${namespace}", job=~"${reportJob}"})',  # noqa: E501
```
```suggestion
        expr='max(s3_cloudserver_quota_buckets_count{namespace="${namespace}", job="${reportJob}"})',  # noqa: E501
```

francoisferrand commented on monitoring/dashboard.py:
```diff
@@ -525,6 +557,172 @@ def top10_errors_by_bucket(title, code):
+    noValue="-",
+    reduceCalc="lastNotNull",
+    targets=[Target(
+        expr='max(s3_cloudserver_quota_accounts_count{namespace="${namespace}", job=~"${reportJob}"})',  # noqa: E501
```
```suggestion
        expr='max(s3_cloudserver_quota_accounts_count{namespace="${namespace}", job="${reportJob}"})',  # noqa: E501
```

francoisferrand commented on monitoring/dashboard.py:
```diff
@@ -525,6 +557,172 @@ def top10_errors_by_bucket(title, code):
+        Threshold("blue", 1, 0.0),
+    ],
+)
+
```
`s3_cloudserver_quota_request_with_metrics_unavailable` is maybe the most important metric, should be in dashboard (generally: if you have an alert, you should be able to monitor the metric...)

williamlardier replied:
Added as a timeserie (middle-top)
![image](https://github.com/scality/cloudserver/assets/91462779/442961f3-519c-4b73-ad3a-93b5abd861f1)

This directly maps with the alert, as we have one if we have >0 values for 5min (maybe it should be less?)

The description of the critical alert was also updated, to be understandable by someone looking at the platform without context

francoisferrand commented on lib/api/apiUtils/quotas/quotaUtils.js:
```diff
@@ -225,17 +246,20 @@ function validateQuotas(request, bucket, account, apiNames, apiMethod, inflight,
         }
         if (actionWithDataDeletion[apiMethod]) {
-            return callback();
+            monitorQuotaEvaluation(apiMethod, type, 200, responseTimeInNs / 1e9);
```
duplicated with call on line 262

francoisferrand commented on lib/api/apiUtils/quotas/quotaUtils.js:
```diff
@@ -225,17 +246,20 @@ function validateQuotas(request, bucket, account, apiNames, apiMethod, inflight,
+            monitorQuotaEvaluation(apiMethod, type, 429, responseTimeInNs / 1e9);
             return callback(errors.QuotaExceeded);
         }
+        monitorQuotaEvaluation(apiMethod, type, 200, responseTimeInNs / 1e9);
```
this computes the metrics for the call to `validateQuotas` ; however, it does not account for the eventual "cleanup" which may happen at the end of failed request...
→ to be "precise", we should store the value here (in the request), always call the cleanup callback, and there also add the cleanup time (if any) before observing the value...
→ otherwise, maybe we should do a second measure if the cleanup is executed.... but then maybe the metric is not correctly named? or maybe the cleanup can be observed via another metric?
→ or maybe this is correct, as we are really only interested in the actual _evaluation_ part, and not the "total cost of quota" ? That may be fine, but should be made clear in the comments (or description?) of the metric

Seragonia committed:
CLDSRV-520: monitor quota cleanup

The finalization of quota logic will always be executed. Some tests
are added to ensure the inflights are only cleaned when they are
enabled, and an error happened in the API.
In any case, this ensures we monitor quotas in a single place,
for each of the executed action, and compute correctly the total
duration of the quota impact on the API.

Seragonia committed:
CLDSRV-520: deduplicate quota logic

Seragonia committed:
CLDSRV-520: generic quota retrieval latency dashboard

Seragonia committed:
CLDSRV-520: bump cloudserver version

francoisferrand commented on lib/utilities/monitoringHandler.js:
```diff
@@ -64,6 +65,49 @@ const httpResponseSizeBytes = new client.Summary({
+    });
+
+    requestWithQuotaMetricsUnavailable = new client.Counter({
+        name: 's3_cloudserver_quota_request_with_metrics_unavailable',
```
i guess just`s3_cloudserver_quota_unavailable_count` is enough ?
i.e. the number of "events" where quota were not available...

francoisferrand commented on lib/utilities/monitoringHandler.js:
```diff
@@ -143,6 +187,10 @@ function crrCacheToProm(crrResults) {
             numberOfObjects.set(crrResults.getObjectCount.objects || 0);
         }
+        if (config.isQuotaEnabled) {
+            numberOfBucketQuota.set(crrResults?.getObjectCount?.bucketWithQuotaCount || 0);
+            numberOfAccountQuota.set(crrResults?.getVaultReport?.accountWithQuota || 0);
```
nit:
```suggestion
            bucketsWithQuota.set(crrResults?.getObjectCount?.bucketWithQuotaCount || 0);
            accountsWithQuota.set(crrResults?.getVaultReport?.accountWithQuota || 0);
```

francoisferrand commented on monitoring/alerts.yaml:
```diff
@@ -132,3 +135,42 @@ groups:
+  # being unavailable
+  - alert: QuotaUnavailable
+    expr: |
+      sum(rate(s3_cloudserver_quota_request_with_metrics_unavailable{namespace="${namespace}",service="${service}"}[2m]))
```
may be more relevant with `increase()`, to limit rounding of values close to 0...

francoisferrand commented on monitoring/dashboard.py:
```diff
@@ -366,6 +366,32 @@ def average_latency_target(title, action="", by=""):
+    by = " by (" + by + ")" if by else ""
+    return Target(
+        expr="\n".join([
+            'sum(rate(s3_cloudserver_quota_evaluation_duration_seconds_sum{namespace="${namespace}", job=~"$job"' + extra + "}[$__rate_interval]))" + by,  # noqa: E501
```
```suggestion
            'sum(rate(s3_cloudserver_quota_evaluation_duration_seconds_sum{namespace="${namespace}", job="${job}"' + extra + "}[$__rate_interval]))" + by,  # noqa: E501
```

francoisferrand commented on monitoring/dashboard.py:
```diff
@@ -366,6 +366,32 @@ def average_latency_target(title, action="", by=""):
+        expr="\n".join([
+            'sum(rate(s3_cloudserver_quota_evaluation_duration_seconds_sum{namespace="${namespace}", job=~"$job"' + extra + "}[$__rate_interval]))" + by,  # noqa: E501
+            "   /",
+            'sum(rate(s3_cloudserver_quota_evaluation_duration_seconds_count{namespace="${namespace}", job=~"$job"' + extra + "}[$__rate_interval]))" + by,  # noqa: E501,
```
```suggestion
            'sum(rate(s3_cloudserver_quota_evaluation_duration_seconds_count{namespace="${namespace}", job="${job}"' + extra + "}[$__rate_interval]))" + by,  # noqa: E501,
```

francoisferrand commented on monitoring/dashboard.py:
```diff
@@ -366,6 +366,32 @@ def average_latency_target(title, action="", by=""):
+    extra = ', code=' + code if code else ""
+    by = " by (" + by + ")" if by else ""
+    return "\n".join([
+        'sum(rate(s3_cloudserver_quota_metrics_retrieval_duration_seconds_sum{namespace="${namespace}", job=~"$job"' + extra + "}[$__rate_interval]))" + by,  # noqa: E501
```
```suggestion
        'sum(rate(s3_cloudserver_quota_metrics_retrieval_duration_seconds_sum{namespace="${namespace}", job="${job}"' + extra + "}[$__rate_interval]))" + by,  # noqa: E501
```

francoisferrand commented on monitoring/dashboard.py:
```diff
@@ -366,6 +366,32 @@ def average_latency_target(title, action="", by=""):
+    return "\n".join([
+        'sum(rate(s3_cloudserver_quota_metrics_retrieval_duration_seconds_sum{namespace="${namespace}", job=~"$job"' + extra + "}[$__rate_interval]))" + by,  # noqa: E501
+        "   /",
+        'sum(rate(s3_cloudserver_quota_metrics_retrieval_duration_seconds_count{namespace="${namespace}", job=~"$job"' + extra + "}[$__rate_interval]))" + by,  # noqa: E501,
```
```suggestion
        'sum(rate(s3_cloudserver_quota_metrics_retrieval_duration_seconds_count{namespace="${namespace}", job="${job}"' + extra + "}[$__rate_interval]))" + by,  # noqa: E501,
```

francoisferrand commented on monitoring/dashboard.py:
```diff
@@ -525,6 +551,189 @@ def top10_errors_by_bucket(title, code):
+    fillOpacity=30,
+    unit=UNITS.PERCENT_FORMAT,
+    targets=[Target(
+        expr='avg(avg_over_time(s3_cloudserver_quota_utilization_service_available{namespace="${namespace}",job=~"$job"}[$__rate_interval])) * 100',  # noqa: E501
```
```suggestion
        expr='avg(avg_over_time(s3_cloudserver_quota_utilization_service_available{namespace="${namespace}",job="${job}"}[$__rate_interval])) * 100',  # noqa: E501
```

francoisferrand commented on monitoring/dashboard.py:
```diff
@@ -525,6 +551,189 @@ def top10_errors_by_bucket(title, code):
+    lineInterpolation="smooth",
+    unit=UNITS.OPS_PER_SEC,
+    targets=[Target(
+        expr='sum(rate(s3_cloudserver_quota_evaluation_duration_seconds_count{namespace="${namespace}", code=~"2..", job=~"$job"}[$__rate_interval]))',  # noqa: E501
```
```suggestion
        expr='sum(rate(s3_cloudserver_quota_evaluation_duration_seconds_count{namespace="${namespace}", code=~"2..", job="${job}"}[$__rate_interval]))',  # noqa: E501
```

francoisferrand commented on monitoring/dashboard.py:
```diff
@@ -525,6 +551,189 @@ def top10_errors_by_bucket(title, code):
+        expr='sum(rate(s3_cloudserver_quota_evaluation_duration_seconds_count{namespace="${namespace}", code=~"2..", job=~"$job"}[$__rate_interval]))',  # noqa: E501
+        legendFormat="Success",
+    ), Target(
+        expr='sum(rate(s3_cloudserver_quota_evaluation_duration_seconds_count{namespace="${namespace}", code="429", job=~"$job"}[$__rate_interval]))',  # noqa: E501
```
```suggestion
        expr='sum(rate(s3_cloudserver_quota_evaluation_duration_seconds_count{namespace="${namespace}", code="429", job="${job}"}[$__rate_interval]))',  # noqa: E501
```

francoisferrand commented on monitoring/dashboard.py:
```diff
@@ -525,6 +551,189 @@ def top10_errors_by_bucket(title, code):
+    unit=UNITS.OPS_PER_SEC,
+    targets=[
+        Target(
+            expr='sum(rate(s3_cloudserver_quota_evaluation_duration_seconds_count{namespace="${namespace}", job=~"$job"}[$__rate_interval])) by(action)',  # noqa: E501
```
```suggestion
            expr='sum(rate(s3_cloudserver_quota_evaluation_duration_seconds_count{namespace="${namespace}", job="${job}"}[$__rate_interval])) by(action)',  # noqa: E501
```

francoisferrand commented on monitoring/dashboard.py:
```diff
@@ -525,6 +551,189 @@ def top10_errors_by_bucket(title, code):
+    yAxis=YAxis(format=UNITS.SECONDS),
+    color=HeatmapColor(mode="opacity"),
+    targets=[Target(
+        expr='sum by(le) (increase(s3_cloudserver_quota_evaluation_duration_seconds_bucket{namespace="${namespace}", job=~"$job"}[$__rate_interval]))',  # noqa: E501
```
```suggestion
        expr='sum by(le) (increase(s3_cloudserver_quota_evaluation_duration_seconds_bucket{namespace="${namespace}", job="${job}"}[$__rate_interval]))',  # noqa: E501
```

francoisferrand commented on monitoring/dashboard.py:
```diff
@@ -525,6 +551,189 @@ def top10_errors_by_bucket(title, code):
+    unit=UNITS.OPS_PER_SEC,
+    legendDisplayMode="hidden",
+    targets=[Target(
+        expr='sum(rate(s3_cloudserver_quota_request_with_metrics_unavailable{namespace="${namespace}", job=~"$job"}[$__rate_interval]))',  # noqa: E501
```
```suggestion
        expr='sum(rate(s3_cloudserver_quota_request_with_metrics_unavailable{namespace="${namespace}", job="${job}"}[$__rate_interval]))',  # noqa: E501
```

bert-e commented:
# Hello williamlardier,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/cloudserver/doc/user).





<details>
  <summary><b>Available options</b></summary>

  name   | description  | privileged   | authored
  ------ | ------------ | ------------ |----------
  `/after_pull_request` | Wait for the given pull request id to be merged before continuing with the current one. |  | 
  `/bypass_author_approval` | Bypass the pull request author's approval |  :star:  | 
  `/bypass_build_status` | Bypass the build and test status |  :star:  | 
  `/bypass_commit_size` | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
  `/bypass_incompatible_branch` | Bypass the check on the source branch prefix |  :star:  | 
  `/bypass_jira_check` | Bypass the Jira issue check |  :star:  | 
  `/bypass_peer_approval` | Bypass the pull request peers' approval |  :star:  | 
  `/bypass_leader_approval` | Bypass the pull request leaders' approval |  :star:  | 
  `/approve` | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
  `/create_pull_requests` | Allow the creation of integration pull requests. |  | 
  `/create_integration_branches` | Allow the creation of integration branches. |  | 
  `/no_octopus` | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
  `/unanimity` | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
  `/wait` | Instruct Bert-E not to run until further notice. |  | 
  

</details>



<details>
  <summary><b>Available commands</b></summary>

  name   | description  | privileged
  ------ | ------------ | ------------
  `/help` | Print Bert-E's manual in the pull request. | 
  `/status` | Print Bert-E's current status in the pull request ```TBA``` | 
  `/clear` | Remove all comments from Bert-E from the history ```TBA``` | 
  `/retry` | Re-start a fresh build ```TBA``` | 
  `/build` | Re-start a fresh build ```TBA``` | 
  `/force_reset` | Delete integration branches & pull requests, and restart merge process from the beginning. | 
  `/reset` | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
  
</details>




*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-520 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.8.21`


Please check the `Fix Version/s` of CLDSRV-520, or the target
branch of this pull request.

francoisferrand commented on monitoring/dashboard.py:
```diff
@@ -525,6 +551,189 @@ def top10_errors_by_bucket(title, code):
+        average_quota_latency_target(
+            title="Account & bucket, exceeded",
+            type='"bucket+account"',
+            code='"429"'),
```
can be deduplicated into just 2 target definitions (`code~"2.."` and `code="429"`), by using queries aggregated `by(type)` with a dynamic legend (`{{ type }}`)

williamlardier replied:
Updated in https://github.com/scality/cloudserver/pull/5556/commits/293d3afeaaf88d8f489fe70ad4b0bf118c07c14c
I test it on the platform, here is the visual:

![image](https://github.com/scality/cloudserver/assets/91462779/8e55879f-de9b-440d-a365-e96d192a354a)

francoisferrand commented on lib/api/apiUtils/quotas/quotaUtils.js:
```diff
@@ -64,11 +65,124 @@ function isMetricStale(metric, resourceType, resourceName, action, inflight, log
+ * @param {number} inflight - The number of inflight requests.
+ * @param {number} inflightForCheck - The number of inflight requests for checking quotas.
+ * @param {string} action - The action being performed.
+ * @param {boolean} cleanUp - Flag indicating whether to ignore quotas.
```
i don't think this flag is even necessary: we could just rely on the fact that inflight will be `< 0` in that case, and just ignore the quota check result: it will be the same number of request anyway...

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

francoisferrand commented on lib/api/apiUtils/quotas/quotaUtils.js:
```diff
@@ -64,11 +65,124 @@ function isMetricStale(metric, resourceType, resourceName, action, inflight, log
 }
 
+/**
+ * Evaluates quotas for a bucket and an account.
```
nit:
```suggestion
 * Evaluates quotas for a bucket and an account and update inflight count.
```

francoisferrand commented on lib/api/apiUtils/quotas/quotaUtils.js:
```diff
@@ -118,113 +247,51 @@ function validateQuotas(request, bucket, account, apiNames, apiMethod, inflight,
+                    const code = (bucketQuotaExceeded || accountQuotaExceeded) ? 429 : 200;
+                    const quotaCleanUpStartTime = process.hrtime.bigint();
+                    // Quotas are cleaned only in case of error in the API
+                    if (errorFromAPI) {
+                        return _evaluateQuotas(bucketQuota, accountQuota, bucket, account, _inflights,
+                            null, apiName, true, log, () => {
+                                monitorQuotaEvaluationDuration(apiMethod, type, code, quotaEvaluationDuration +
+                                    Number(process.hrtime.bigint() - quotaCleanUpStartTime));
+                                return _done();
+                            });
+                    }
+                    monitorQuotaEvaluationDuration(apiMethod, type, code, quotaEvaluationDuration);
                     return _done();
```
use waterflow to simplify code:

```suggestion
                    async.waterflow(
                        cb => {
                            if (errorFromAPI) {
                                return _evaluateQuotas(bucketQuota,      accountQuota, bucket, account, _inflights,
                            null, apiName, true, log, cb);
                            }
                            return cb();
                        },
                        () => {
                            monitorQuotaEvaluationDuration(apiMethod, type, code, quotaEvaluationDuration +
                                Number(process.hrtime.bigint() - quotaCleanUpStartTime));
                            return _done();
                        }
                    )
```

francoisferrand commented on lib/quotas/scuba/wrapper.js:
```diff
@@ -60,7 +63,22 @@ class ScubaClientImpl extends ScubaClient {
+        const requestStartTime = process.hrtime.bigint();
+        return this._getLatestMetricsCallback(metricsClass, resourceName, options, body, (err, data) => {
+            const responseTimeInNs = Number(process.hrtime.bigint() - requestStartTime);
+            if (err) {
+                monitoring.utilizationMetricsRetrievalDuration.labels({
+                    code: err.statusCode || 500,
+                    class: metricsClass,
+                }).observe(responseTimeInNs / 1e9);
+                return callback(err);
+            }
+            monitoring.utilizationMetricsRetrievalDuration.labels({
+                code: 200,
+                class: metricsClass,
+            }).observe(responseTimeInNs / 1e9);
+            return callback(null, data);
```
```suggestion
            monitoring.utilizationMetricsRetrievalDuration.labels({
                code: err ? (err.statusCode || 500) : 200,
                class: metricsClass,
            }).observe(responseTimeInNs / 1e9);
            return callback(err, data);
```

francoisferrand reviewed (approved):
LGTM, just minor cleanups

ghost commented:
/approve

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.8`



The following branches have **NOT** changed:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.6`
* `development/8.7`



Please check the status of the associated issue CLDSRV-520.

Goodbye williamlardier.




*The following options are set:* **approve**
